File: backend/src/routes/projects.ts
import express from "express"
import { z } from "zod"
import { getProjects, getProject, createProject, updateProject, deleteProject } from "../services/db/project.service.js"
import { updateSlot, updateSlots } from "../services/db/slot.service.js"

const router = express.Router()

/**
 * GET /api/projects
 * プロジェクト一覧を取得
 */
router.get("/", async (req, res) => {
  try {
    const projects = await getProjects()
    res.json({
      success: true,
      data: projects,
    })
  } catch (error) {
    console.error("Error fetching projects:", error)
    res.status(500).json({
      success: false,
      error: "Failed to fetch projects",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

/**
 * GET /api/projects/:id
 * プロジェクトを取得
 */
router.get("/:id", async (req, res) => {
  try {
    const { id } = req.params
    const project = await getProject(id)

    if (!project) {
      return res.status(404).json({
        success: false,
        error: "Project not found",
      })
    }

    res.json({
      success: true,
      data: project,
    })
  } catch (error) {
    console.error("Error fetching project:", error)
    res.status(500).json({
      success: false,
      error: "Failed to fetch project",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

/**
 * POST /api/projects
 * プロジェクトを作成
 */
router.post("/", async (req, res) => {
  try {
    const schema = z.object({
      name: z.string().min(1),
      description: z.string().optional().default(""),
      templateId: z.string().uuid(),
      templateName: z.string().min(1),
    })

    const data = schema.parse(req.body)
    const project = await createProject(data)

    res.json({
      success: true,
      data: project,
    })
  } catch (error) {
    console.error("Error creating project:", error)
    if (error instanceof z.ZodError) {
      res.status(400).json({
        success: false,
        error: "Invalid request data",
        details: error.errors,
      })
    } else {
      res.status(500).json({
        success: false,
        error: "Failed to create project",
        message: error instanceof Error ? error.message : "Unknown error",
      })
    }
  }
})

/**
 * PATCH /api/projects/:id
 * プロジェクトを更新
 */
router.patch("/:id", async (req, res) => {
  try {
    const { id } = req.params
    const schema = z.object({
      name: z.string().min(1).optional(),
      description: z.string().optional(),
      status: z.enum(["draft", "in-progress", "completed"]).optional(),
    })

    const data = schema.parse(req.body)
    const project = await updateProject(id, data)

    res.json({
      success: true,
      data: project,
    })
  } catch (error) {
    console.error("Error updating project:", error)
    if (error instanceof z.ZodError) {
      res.status(400).json({
        success: false,
        error: "Invalid request data",
        details: error.errors,
      })
    } else {
      res.status(500).json({
        success: false,
        error: "Failed to update project",
        message: error instanceof Error ? error.message : "Unknown error",
      })
    }
  }
})

/**
 * DELETE /api/projects/:id
 * プロジェクトを削除
 */
router.delete("/:id", async (req, res) => {
  try {
    const { id } = req.params
    await deleteProject(id)

    res.json({
      success: true,
      message: "Project deleted successfully",
    })
  } catch (error) {
    console.error("Error deleting project:", error)
    res.status(500).json({
      success: false,
      error: "Failed to delete project",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

/**
 * PATCH /api/projects/:id/slots/:slotId
 * スロットを更新
 */
router.patch("/:id/slots/:slotId", async (req, res) => {
  try {
    const { id, slotId } = req.params
    const schema = z.object({
      value: z.string().optional(),
      status: z.enum(["empty", "draft", "filled", "approved"]).optional(),
      linkedStoryId: z.string().uuid().nullable().optional(),
    })

    const data = schema.parse(req.body)
    const slot = await updateSlot(id, slotId, data)

    res.json({
      success: true,
      data: slot,
    })
  } catch (error) {
    console.error("Error updating slot:", error)
    if (error instanceof z.ZodError) {
      res.status(400).json({
        success: false,
        error: "Invalid request data",
        details: error.errors,
      })
    } else {
      res.status(500).json({
        success: false,
        error: "Failed to update slot",
        message: error instanceof Error ? error.message : "Unknown error",
      })
    }
  }
})

/**
 * PATCH /api/projects/:id/slots
 * スロットを一括更新
 */
router.patch("/:id/slots", async (req, res) => {
  try {
    const { id } = req.params
    const schema = z.object({
      updates: z.array(
        z.object({
          slotId: z.string().uuid(),
          value: z.string().optional(),
          status: z.enum(["empty", "draft", "filled", "approved"]).optional(),
        })
      ),
    })

    const { updates } = schema.parse(req.body)
    const slots = await updateSlots(id, updates)

    res.json({
      success: true,
      data: slots,
    })
  } catch (error) {
    console.error("Error updating slots:", error)
    if (error instanceof z.ZodError) {
      res.status(400).json({
        success: false,
        error: "Invalid request data",
        details: error.errors,
      })
    } else {
      res.status(500).json({
        success: false,
        error: "Failed to update slots",
        message: error instanceof Error ? error.message : "Unknown error",
      })
    }
  }
})

export default router

File: backend/app/routers/projects.py
from fastapi import APIRouter, HTTPException, Depends, Path
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.database import get_db
from app.services.db.project_service import (
    get_projects,
    get_project,
    create_project,
    update_project,
    delete_project,
)
from app.services.db.slot_service import update_slot, update_slots
from typing import Optional, List

router = APIRouter(prefix="/api/projects", tags=["projects"])


class CreateProjectRequest(BaseModel):
    name: str
    description: str = ""
    template_id: str
    template_name: str


class UpdateProjectRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None


class UpdateSlotRequest(BaseModel):
    value: Optional[str] = None
    status: Optional[str] = None
    linked_story_id: Optional[str] = None


class UpdateSlotsRequest(BaseModel):
    updates: List[dict]


@router.get("")
async def list_projects(db: Session = Depends(get_db)):
    """プロジェクト一覧を取得"""
    try:
        projects = get_projects(db)
        return {
            "success": True,
            "data": [project.dict() for project in projects],
        }
    except Exception as e:
        print(f"Error fetching projects: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch projects: {str(e)}",
        )


@router.get("/{project_id}")
async def get_project_by_id(
    project_id: str = Path(...),
    db: Session = Depends(get_db),
):
    """プロジェクトを取得"""
    try:
        project = get_project(db, project_id)
        
        if not project:
            raise HTTPException(status_code=404, detail="Project not found")
        
        return {
            "success": True,
            "data": project.dict(),
        }
    except HTTPException:
        raise
    except Exception as e:
        print(f"Error fetching project: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch project: {str(e)}",
        )


@router.post("")
async def create_new_project(
    request: CreateProjectRequest,
    db: Session = Depends(get_db),
):
    """プロジェクトを作成"""
    try:
        project = create_project(
            db,
            name=request.name,
            description=request.description,
            template_id=request.template_id,
            template_name=request.template_name,
        )
        
        return {
            "success": True,
            "data": project.dict(),
        }
    except Exception as e:
        print(f"Error creating project: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to create project: {str(e)}",
        )


@router.patch("/{project_id}")
async def update_project_by_id(
    project_id: str = Path(...),
    request: UpdateProjectRequest = ...,
    db: Session = Depends(get_db),
):
    """プロジェクトを更新"""
    try:
        project = update_project(
            db,
            project_id,
            name=request.name,
            description=request.description,
            status=request.status,
        )
        
        return {
            "success": True,
            "data": project.dict(),
        }
    except Exception as e:
        print(f"Error updating project: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update project: {str(e)}",
        )


@router.delete("/{project_id}")
async def delete_project_by_id(
    project_id: str = Path(...),
    db: Session = Depends(get_db),
):
    """プロジェクトを削除"""
    try:
        delete_project(db, project_id)
        
        return {
            "success": True,
            "message": "Project deleted successfully",
        }
    except Exception as e:
        print(f"Error deleting project: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to delete project: {str(e)}",
        )


@router.patch("/{project_id}/slots/{slot_id}")
async def update_slot_by_id(
    project_id: str = Path(...),
    slot_id: str = Path(...),
    request: UpdateSlotRequest = ...,
    db: Session = Depends(get_db),
):
    """スロットを更新"""
    try:
        slot = update_slot(
            db,
            project_id,
            slot_id,
            value=request.value,
            status=request.status,
            linked_story_id=request.linked_story_id,
        )
        
        return {
            "success": True,
            "data": slot.dict(),
        }
    except Exception as e:
        print(f"Error updating slot: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update slot: {str(e)}",
        )


@router.patch("/{project_id}/slots")
async def update_slots_bulk(
    project_id: str = Path(...),
    request: UpdateSlotsRequest = ...,
    db: Session = Depends(get_db),
):
    """スロットを一括更新"""
    try:
        slots = update_slots(db, project_id, request.updates)
        
        return {
            "success": True,
            "data": [slot.dict() for slot in slots],
        }
    except Exception as e:
        print(f"Error updating slots: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update slots: {str(e)}",
        )

